### PR TITLE
Bybit fetch balance - cross margin

### PR DESCRIPTION
**Pull request by samgermain**
# Testing

```
# Swap                                                                # Cross Margin                                                                             # Spot

sam@Sams-Mac-mini ccxt % bybit fetchBalance '{"type": "swap"}'        % bybit fetchBalance '{"marginMode": "cross"}'                                      % bybit fetchBalance                          
2022-07-29T20:52:42.574Z                                              2022-07-29T20:49:21.982Z                                                            2022-07-29T20:50:03.399Z
Node.js: v18.4.0                                                      Node.js: v18.4.0                                                                    Node.js: v18.4.0
CCXT v1.91.42                                                         CCXT v1.91.42                                                                       CCXT v1.91.42
bybit.fetchBalance ([object Object])                                  bybit.fetchBalance ([object Object])                                                bybit.fetchBalance ()
2022-07-29T20:52:44.586Z iteration 0 passed in 269 ms                 2022-07-29T20:49:24.958Z iteration 0 passed in 270 ms                               2022-07-29T20:50:05.442Z iteration 0 passed in 277 ms
                                                                                                                                                          
{                                                                     {                                                                                   {
  info: {                                                               info: {                                                                             info: {
    ret_code: '0',                                                        ret_code: '0',                                                                      ret_code: '0',
    ret_msg: 'OK',                                                        ret_msg: '',                                                                        ret_msg: '',
    ext_code: '',                                                         ext_code: null,                                                                     ext_code: null,
    ext_info: '',                                                         ext_info: null,                                                                     ext_info: null,
    result: {                                                             result: {                                                                           result: {
      ADA: {                                                                status: '2',                                                                        balances: [
        equity: '0',                                                        riskRate: '0.6388',                                                                   {
        available_balance: '0',                                             acctBalanceSum: '0.006655298910667173',                                                 coin: 'BTC',
        used_margin: '0',                                                   debtBalanceSum: '0.00425167279005171',                                                  coinId: 'BTC',
        order_margin: '0',                                                  loanAccountList: [                                                                      coinName: 'BTC',
        position_margin: '0',                                                 {                                                                                     total: '0.000000401',
        occ_closing_fee: '0',                                                   tokenId: 'BTC',                                                                     free: '0.000000401',
        occ_funding_fee: '0',                                                   total: '0.000000401',                                                               locked: '0'
        wallet_balance: '0',                                                    locked: '0',                                                                      },
        realised_pnl: '0',                                                      loan: '0',                                                                        {
        unrealised_pnl: '0',                                                    interest: '0',                                                                      coin: 'MANA',
        cum_realised_pnl: '0',                                                  free: '0.000000401'                                                                 coinId: 'MANA',
        given_cash: '0',                                                      },                                                                                    coinName: 'MANA',
        service_cash: '0'                                                     {                                                                                     total: '0.00617',
      },                                                                        tokenId: 'XRP',                                                                     free: '0.00617',
      BIT: {                                                                    total: '432.10746',                                                                 locked: '0'
        equity: '0',                                                            locked: '0',                                                                      },
        available_balance: '0',                                                 loan: '0',                                                                        {
        used_margin: '0',                                                       interest: '0',                                                                      coin: 'USDT',
        order_margin: '0',                                                      free: '432.10746'                                                                   coinId: 'USDT',
        position_margin: '0',                                                 },                                                                                    coinName: 'USDT',
        occ_closing_fee: '0',                                                 {                                                                                     total: '0.00261595736',
        occ_funding_fee: '0',                                                   tokenId: 'USDT',                                                                    free: '0.00261595736',
        wallet_balance: '0',                                                    total: '0.00261595736',                                                             locked: '0'
        realised_pnl: '0',                                                      locked: '0',                                                                      },
        unrealised_pnl: '0',                                                    loan: '101.8965',                                                                 {
        cum_realised_pnl: '0',                                                  interest: '0.0004931384',                                                           coin: 'XRP',
        given_cash: '0',                                                        free: '0.00261595736'                                                               coinId: 'XRP',
        service_cash: '0'                                                     }                                                                                     coinName: 'XRP',
      },                                                                    ]                                                                                       total: '432.10746',
      BTC: {                                                              }                                                                                         free: '432.10746',
        equity: '0',                                                    },                                                                                          locked: '0'
        available_balance: '0',                                         BTC: { free: 4.01e-7, used: 0, total: 4.01e-7 },                                          }
        used_margin: '0',                                               XRP: { free: 432.10746, used: 0, total: 432.10746 },                                    ]
        order_margin: '0',                                              USDT: { free: 0.00261595736, used: 101.8969931384, total: 0.00261595736 },            }
        position_margin: '0',                                           free: { BTC: 4.01e-7, XRP: 432.10746, USDT: 0.00261595736 },                        },
        occ_closing_fee: '0',                                           used: { BTC: 0, XRP: 0, USDT: 101.8969931384 },                                     BTC: { free: 4.01e-7, used: 0, total: 4.01e-7 },
        occ_funding_fee: '0',                                           total: { BTC: 4.01e-7, XRP: 432.10746, USDT: 0.00261595736 }                        MANA: { free: 0.00617, used: 0, total: 0.00617 },
        wallet_balance: '0',                                          }                                                                                     USDT: { free: 0.00261595736, used: 0, total: 0.00261595736 },
        realised_pnl: '0',                                            2022-07-29T20:49:24.958Z iteration 1 passed in 270 ms                                 XRP: { free: 432.10746, used: 0, total: 432.10746 },
        unrealised_pnl: '0',                                                                                                                                free: { BTC: 4.01e-7, MANA: 0.00617, USDT: 0.00261595736, XRP: 432.10746 },
        cum_realised_pnl: '0',                                                                                                                              used: { BTC: 0, MANA: 0, USDT: 0, XRP: 0 },
        given_cash: '0',                                                                                                                                    total: { BTC: 4.01e-7, MANA: 0.00617, USDT: 0.00261595736, XRP: 432.10746 }
        service_cash: '0'                                                                                                                                 }
      },                                                                                                                                                  2022-07-29T20:50:05.442Z iteration 1 passed in 277 ms
      DOT: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      },
      EOS: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      },
      ETH: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      },
      LTC: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      },
      LUNA: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      },
      MANA: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      },
      SOL: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      },
      USDT: {
        equity: '0.00700662',
        available_balance: '0.00700662',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0.00700662',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '-197.09369338',
        given_cash: '0',
        service_cash: '0'
      },
      XRP: {
        equity: '0',
        available_balance: '0',
        used_margin: '0',
        order_margin: '0',
        position_margin: '0',
        occ_closing_fee: '0',
        occ_funding_fee: '0',
        wallet_balance: '0',
        realised_pnl: '0',
        unrealised_pnl: '0',
        cum_realised_pnl: '0',
        given_cash: '0',
        service_cash: '0'
      }
    },
    time_now: '1659127964.694827',
    rate_limit_status: '119',
    rate_limit_reset_ms: '1659127964686',
    rate_limit: '120'
  },
  ADA: { free: 0, used: 0, total: 0 },
  BIT: { free: 0, used: 0, total: 0 },
  BTC: { free: 0, used: 0, total: 0 },
  DOT: { free: 0, used: 0, total: 0 },
  EOS: { free: 0, used: 0, total: 0 },
  ETH: { free: 0, used: 0, total: 0 },
  LTC: { free: 0, used: 0, total: 0 },
  LUNA: { free: 0, used: 0, total: 0 },
  MANA: { free: 0, used: 0, total: 0 },
  SOL: { free: 0, used: 0, total: 0 },
  USDT: { free: 0.00700662, used: 0, total: 0.00700662 },
  XRP: { free: 0, used: 0, total: 0 },
  free: {
    ADA: 0,
    BIT: 0,
    BTC: 0,
    DOT: 0,
    EOS: 0,
    ETH: 0,
    LTC: 0,
    LUNA: 0,
    MANA: 0,
    SOL: 0,
    USDT: 0.00700662,
    XRP: 0
  },
  used: {
    ADA: 0,
    BIT: 0,
    BTC: 0,
    DOT: 0,
    EOS: 0,
    ETH: 0,
    LTC: 0,
    LUNA: 0,
    MANA: 0,
    SOL: 0,
    USDT: 0,
    XRP: 0
  },
  total: {
    ADA: 0,
    BIT: 0,
    BTC: 0,
    DOT: 0,
    EOS: 0,
    ETH: 0,
    LTC: 0,
    LUNA: 0,
    MANA: 0,
    SOL: 0,
    USDT: 0.00700662,
    XRP: 0
  }
}
2022-07-29T20:52:44.586Z iteration 1 passed in 269 ms
```

### CODE REVIEW

File: js/bybit.js [hunk 1]
```diff
@@ -2190,16 +2226,20 @@ module.exports = class bybit extends Exchange {
...
                 const account = this.account ();
-                account['free'] = this.safeString (balance, 'availableBalance');
-                account['used'] = this.safeString (balance, 'locked');
+                const locked = this.safeString (balance, 'locked');
+                const loan = this.safeString (balance, 'loan');
+                const interest = this.safeString (balance, 'interest');
+                const owed = Precise.stringAdd (loan, interest);
+                account['free'] = this.safeString2 (balance, 'availableBalance', 'free');
+                account['used'] = Precise.stringAdd (locked, owed);
                 account['total'] = this.safeString (balance, 'total');
```
Comment (carlosmiei) [inline js/bybit.js:2243]: would not be easier/efficient to let `safeBalance` compute `used` by doing `total - free = used`?

File: js/bybit.js [hunk 2]
```diff
@@ -2190,16 +2226,20 @@ module.exports = class bybit extends Exchange {
...
                 const code = this.safeCurrencyCode (currencyId);
                 const account = this.account ();
-                account['free'] = this.safeString (balance, 'availableBalance');
-                account['used'] = this.safeString (balance, 'locked');
+                const locked = this.safeString (balance, 'locked');
+                const loan = this.safeString (balance, 'loan');
+                const interest = this.safeString (balance, 'interest');
+                const owed = Precise.stringAdd (loan, interest);
+                account['free'] = this.safeString2 (balance, 'availableBalance', 'free');
+                account['used'] = Precise.stringAdd (locked, owed);
```
Comment (carlosmiei) [inline js/bybit.js:2242]: @samgermain would not be easier/efficient to let `safeBalance` compute `used` by doing `total - free = used`?

File: js/bybit.js [hunk 3]
```diff
@@ -2233,15 +2273,27 @@ module.exports = class bybit extends Exchange {
...
+        let marginMode = undefined;
         [ type, params ] = this.handleMarketTypeAndParams ('fetchBalance', undefined, params);
+        [ marginMode, params ] = this.handleMarginModeAndParams ('fetchBalance', params);
         let method = undefined;
         if (type === 'spot') {
-            method = 'privateGetSpotV1Account';
+            if (marginMode === 'cross') {
+                method = 'privateGetSpotV1CrossMarginAccountsBalance';
+            } else if (marginMode === 'isolated') {
+                throw new BadRequest (this.id + ' fetchBalance marginMode must be cross if assigned');
```
Comment (carlosmiei) [inline js/bybit.js:2293]: I would suggest replacing `BadRequest` with `NotSupported` and probably changing the message a bit to something like "fetchBalance() does not support isolated margin...." . Currently I don't find it much clear

**Comment by carlosmiei** [comment]
✅ ✅ working fine, minor comments left

```
 p bybit fetchBalance '{"marginMode":"cross"}' --spot  --sandbox          
Python v3.9.7
CCXT v1.91.64
bybit.fetchBalance({'marginMode': 'cross'})
{'BTC': {'free': 0.00740141118, 'total': 0.00740141118, 'used': 0.0},
 'ETH': {'free': 0.07825167, 'total': 0.07825167, 'used': 0.0},
 'USDT': {'free': 29765.57924378754,
          'total': 29765.57924378754,
          'used': 0.000417376},
 'XRP': {'free': 10000.0, 'total': 10000.0, 'used': 0.0},
 'free': {'BTC': 0.00740141118,
          'ETH': 0.07825167,
          'USDT': 29765.57924378754,
          'XRP': 10000.0},
 'info': {'ext_code': None,
          'ext_info': None,
          'result': {'acctBalanceSum': '1.658978739661973027',
                     'debtBalanceSum': '0.000000022311590132',
                     'loanAccountList': [{'free': '0.00740141118',
                                          'interest': '0',
                                          'loan': '0',
                                          'locked': '0',
                                          'tokenId': 'BTC',
                                          'total': '0.00740141118'},
                                         {'free': '10000',
                                          'interest': '0',
                                          'loan': '0',
                                          'locked': '0',
                                          'tokenId': 'XRP',
                                          'total': '10000'},
                                         {'free': '0.07825167',
                                          'interest': '0',
                                          'loan': '0',
                                          'locked': '0',
                                          'tokenId': 'ETH',
                                          'total': '0.07825167'},
                                         {'free': '29765.57924378754',
                                          'interest': '0.000000676',
                                          'loan': '0.0004167',
                                          'locked': '0',
                                          'tokenId': 'USDT',
                                          'total': '29765.57924378754'}],
                     'riskRate': '0',
                     'status': '1'},
          'ret_code': '0',
          'ret_msg': ''},
 'total': {'BTC': 0.00740141118,
           'ETH': 0.07825167,
           'USDT': 29765.57924378754,
           'XRP': 10000.0},
 'used': {'BTC': 0.0, 'ETH': 0.0, 'USDT': 0.000417376, 'XRP': 0.0}}
```
```

(base) ➜  ccxt git:(bybit-fetch-balance-margin) ✗ p bybit fetchBalance --spot  --sandbox
Python v3.9.7
CCXT v1.91.64
bybit.fetchBalance()
{'BTC': {'free': 0.00740141118, 'total': 0.00740141118, 'used': 0.0},
 'EOS': {'free': 2000.0, 'total': 2000.0, 'used': 0.0},
 'ETH': {'free': 0.07825167, 'total': 0.07825167, 'used': 0.0},
 'USDC': {'free': 45000.0, 'total': 45000.0, 'used': 0.0},
 'USDT': {'free': 29765.57924378754, 'total': 29765.57924378754, 'used': 0.0},
 'XRP': {'free': 10000.0, 'total': 10000.0, 'used': 0.0},
 'free': {'BTC': 0.00740141118,
          'EOS': 2000.0,
          'ETH': 0.07825167,
          'USDC': 45000.0,
          'USDT': 29765.57924378754,
          'XRP': 10000.0},
 'info': {'ext_code': None,
          'ext_info': None,
          'result': {'balances': [{'coin': 'BTC',
                                   'coinId': 'BTC',
                                   'coinName': 'BTC',
                                   'free': '0.00740141118',
                                   'locked': '0',
                                   'total': '0.00740141118'},
                                  {'coin': 'EOS',
                                   'coinId': 'EOS',
                                   'coinName': 'EOS',
                                   'free': '2000',
                                   'locked': '0',
                                   'total': '2000'},
                                  {'coin': 'ETH',
                                   'coinId': 'ETH',
                                   'coinName': 'ETH',
                                   'free': '0.07825167',
                                   'locked': '0',
                                   'total': '0.07825167'},
                                  {'coin': 'USDC',
                                   'coinId': 'USDC',
                                   'coinName': 'USDC',
                                   'free': '45000',
                                   'locked': '0',
                                   'total': '45000'},
                                  {'coin': 'USDT',
                                   'coinId': 'USDT',
                                   'coinName': 'USDT',
                                   'free': '29765.57924378754',
                                   'locked': '0',
                                   'total': '29765.57924378754'},
                                  {'coin': 'XRP',
                                   'coinId': 'XRP',
                                   'coinName': 'XRP',
                                   'free': '10000',
                                   'locked': '0',
                                   'total': '10000'}]},
          'ret_code': '0',
          'ret_msg': ''},
 'total': {'BTC': 0.00740141118,
           'EOS': 2000.0,
           'ETH': 0.07825167,
           'USDC': 45000.0,
           'USDT': 29765.57924378754,
           'XRP': 10000.0},
 'used': {'BTC': 0.0,
          'EOS': 0.0,
          'ETH': 0.0,
          'USDC': 0.0,
          'USDT': 0.0,
          'XRP': 0.0}}

**Comment by carlosmiei** [comment]
@samgermain can you check the comments I left? very minor things but would like to hear your thoughts on them before merging

File: js/bybit.js (hunk 2 above)
Comment (samgermain) [reply js/bybit.js:2242]: That wouldn't be correct. 

- The total amount is the amount that the user owns.
- The free amount is the amount unused that is in the users wallet
- The used amount is the amount of currency that is locked up and not available for use, which includes the amount of borrowed currency and the amount of currency in trades

I'm realizing that this method might not be correct either. I'm not sure what the correct method would be. We should probably get the opinion of @kroitor @frosty00

Comment (carlosmiei) [reply js/bybit.js:2242]: @samgermain any update?

Comment (samgermain) [reply js/bybit.js:2242]: @carlosmiei @kroitor @frosty00 

No, I need a second opinion

**Comment by frosty00** [comment]
closing because I have added support for cross margin to bybit v3 fetchBalance

**Pull request closed without merging**